### DISCUSSION TITLE
How do I separate a prompt into "sections" of groups of segments?

### DISCUSSION THREAD

**Discussion by pfmoore** [JanDeDobbeleer/oh-my-posh#4928] [General]
I'm trying to construct a prompt that has various "sections". Each section ends with a powerline ">" symbol, then a gap, then the next section. That's relatively easy, you can do it fine with something like this:

```
{
    "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
    "blocks": [
      {
        "alignment": "left",
        "segments": [
          {"type":"text", "style":"diamond", "leading_diamond":"\ue0d7", "trailing_diamond":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"Part1"},
          {"type":"text", "style":"diamond", "leading_diamond":"\ue0d7", "trailing_diamond":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"A"},
          {"type":"text", "style":"diamond", "trailing_diamond":"\ue0b0", "foreground":"#1c1c1c", "background":"#b925af","template":"B"},
          {"type":"text", "style":"diamond", "trailing_diamond":"\ue0b0", "foreground":"#1c1c1c", "background":"#df74d8","template":"C"},
          {"type":"text", "style":"diamond", "leading_diamond":"\ue0d7", "trailing_diamond":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"Part2"}
        ],
        "type": "prompt"
      }
    ],
    "final_space": true,
    "version": 2
  }
```

(the "leading diamond" is an inverted Powerline diamond, to avoid problems with transparent forground not rendering properly on Windows Terminal). The problem is that if A doesn't display, B merges into Part 1. I tried adding a zero-width text column between part 1 and A/B/C, but then the spacing goes wrong.

The reason I want this sort of layout is that I want to split my prompt into sections: system data, directory, VCS, language versions, times. Within each section there may be zero, one or many sections (a directory could have no code, all Python, or a mix of Python and Javascript, for example).

Is this possible? If it isn't, I can simply design a different layout, but I was hoping to do something like this because it's a bit more readable than an unbroken prompt and better structured than having *every* segment separated from the others.

**Comment by mjcarman** [comment]
If I'm understanding you correctly, you can get the behavior you want by splitting your prompt into three blocks and using powerline style within each block. e.g.

```
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {"type":"text", "style":"powerline", "powerline_symbol":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"Part1"}
      ],
      "type": "prompt"
    },
    {
      "alignment": "left",
      "segments": [
        {"type":"text", "style":"powerline", "powerline_symbol":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"A"},
        {"type":"text", "style":"powerline", "powerline_symbol":"\ue0b0", "foreground":"#1c1c1c", "background":"#b925af","template":"B"},
        {"type":"text", "style":"powerline", "powerline_symbol":"\ue0b0", "foreground":"#1c1c1c", "background":"#df74d8","template":"C"}
      ],
      "type": "prompt"
    },
    {
      "alignment": "left",
      "segments": [
        {"type":"text", "style":"powerline", "powerline_symbol":"\ue0b0", "foreground":"#1c1c1c", "background":"#95378f","template":"Part2"}
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
``` 

Using "diamond" to emulate "powerline" is problematic because if the leading segment isn't displayed the next segment which is rendered won't render with the ">" (or whatever character) you want that range of segments to begin with.

**Reply by pfmoore** [reply]
Thanks. The problem with “powerline” is that it doesn’t render correctly when there is a transparent background before, because Windows Terminal doesn’t support transparent foreground (at least not the way oh my posh does it by setting the foreground and then inverting). It’s a really annoying limitation that makes powerline form useless for split prompts 🙁

**Comment by JanDeDobbeleer** [comment]
Is this what you're trying to achieve?

<img width="418" alt="image" src="https://github.com/JanDeDobbeleer/oh-my-posh/assets/2492783/cf791348-a2ff-48a4-a34e-f427ebfdcc6a">

<details>
<summary>Code</summary>

```json
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "text",
          "style": "powerline",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "alias": "A",
          "template": " A "
        },
        {
          "type": "text",
          "style": "powerline",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#b925af",
          "alias": "B",
          "template": " B "
        },
        {
          "type": "text",
          "style": "powerline",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#df74d8",
          "alias": "C",
          "template": " C "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```
</details>

Or this?

<img width="418" alt="image" src="https://github.com/JanDeDobbeleer/oh-my-posh/assets/2492783/ddfa49cc-87c9-4670-a66b-788cca27000d">

<details>
<summary>Code</summary>

```json
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "alias": "A",
          "template": " A "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#b925af",
          "alias": "B",
          "template": " B "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#df74d8",
          "alias": "C",
          "template": " C "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "<transparent,background>\ue0b0</>",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```
</details>

**Reply by pfmoore** [reply]
The first. And the problem is that the starts of the sections have a black triangle instead of transparent:

![image](https://github.com/JanDeDobbeleer/oh-my-posh/assets/1110419/4913630d-8a67-4450-8da5-741a1f8adbdd)

**Reply by pfmoore** [reply]
The underlying Windows Terminal issue is https://github.com/microsoft/terminal/issues/7014 (which I'm pretty sure I originally found here somewhere, so I'm sure you know about it). But it doesn't look like it's going to be fixed particularly soon, which is why I was looking for workarounds. I recently discovered that Nerd Fonts had added the "inverse powerline" glyphs at https://github.com/ryanoasis/nerd-fonts/pull/1490, which prompted me to look at the "diamond" format. I think the reason it doesn't work is because it doesn't have the adaptive behaviour that the powerline format does.

Maybe there's a possibility of using a variant of powerline that, instead of using the `powerline_symbol` with inverted colours at the start, uses an `inverted_powerline_symbol` with non-inverted colours? That would be a new feature, though, which is why I wanted to see if there was a way to get things working with the existing features before I suggested it.

**Reply by JanDeDobbeleer** [reply]
The problem with the inverted symbol is that it doesn't align correctly with the rest of the prompt, although that might have changed in recent versions. I can validate that.

**Reply by pfmoore** [reply]
Hmm, does it not? I'm using it in some test prompts, and it seemed OK to me:

![image](https://github.com/JanDeDobbeleer/oh-my-posh/assets/1110419/ee835f00-bd23-4730-a2c0-1362140a92af)

Hmm, do you mean the cut off end of the triangle on the right of the path and execution time segments? I guess that's not ideal, but presumably the nerd font glyph could be fixed - and even so, it's better than black where the prompt should be transparent 🙂

If that is the issue you mean, let me know and I'll post it to the nerd font tracker.

**Reply by JanDeDobbeleer** [reply]
@pfmoore this one works when you never have all segments disabled:

<details>
<summary>Code</summary>

```json
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "alias": "A",
          "template": " A "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#b925af",
          "alias": "B",
          "template": " B "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#df74d8",
          "alias": "C",
          "template": " C "
        },
        {
          "type": "text",
          "style": "plain",
          "template": "<parentBackground,transparent>\ue0b0</>"
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```
</details>

**Reply by JanDeDobbeleer** [reply]
@pfmoore actually, this seems to work

![image](https://github.com/JanDeDobbeleer/oh-my-posh/assets/2492783/abfe717a-3460-4cf0-ac4d-d1d01ce64d60)

<details>
<summary>Code</summary>

```json
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "alias": "A",
          "template": " A "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#b925af",
          "alias": "B",
          "template": " B "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "foreground": "#1c1c1c",
          "background": "#df74d8",
          "alias": "C",
          "template": " C "
        },
        {
          "type": "text",
          "style": "plain",
          "template": "{{ if gt (len .Segments) 1 }}<parentBackground,transparent>\ue0b0</>{{ end }}"
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```
</details>

**Reply by pfmoore** [reply]
Oh neat! I wondered about having some sort of "final item that just had the trailing diamond" segment, but I couldn't work out how the `.Segments` thing worked.

I'm still a bit unclear - is `(len .Segments)` the number of segments *before* the current one? Or to put it another way, does `.Segments` only contain the segments that have already been processed at this point? I couldn't find much in the documentation about it.

**Reply by JanDeDobbeleer** [reply]
@pfmoore it contains every segment of that block which then should in this case imply `if gt 2`, but somehow the count is off. I need to check that.

**Reply by pfmoore** [reply]
Ah, OK. So it would fail to work if there were *two* multi-segment parts like A>B>C. Maybe that could be fixed by using multiple blocks, I'll look into that a bit further.

I really appreciate all the help here, and I don't want to sound negative, but it is all a bit complicated for something that's basically a workaround for the Windows Terminal transparent foreground bug. Is there any mileage in the idea of an `inverted_powerline_symbol` option, if I could get them to fix the nerd font issue?

**Reply by JanDeDobbeleer** [reply]
@pfmoore the last addition to nerd fonts (PR you linked) seems to work perfectly.

**Reply by pfmoore** [reply]
Oh that's great.

**Reply by JanDeDobbeleer** [reply]
https://github.com/JanDeDobbeleer/oh-my-posh/pull/4955

<details>
<summary>Code</summary>

```json
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "text",
          "style": "powerline",
          "leading_powerline_symbol": "\ue0d7",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "alias": "A",
          "template": " A "
        },
        {
          "type": "text",
          "style": "powerline",
          "leading_powerline_symbol": "\ue0d7",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#b925af",
          "alias": "B",
          "template": " B "
        },
        {
          "type": "text",
          "style": "powerline",
          "leading_powerline_symbol": "\ue0d7",
          "powerline_symbol": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#df74d8",
          "alias": "C",
          "template": " C "
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```

</details>

**Comment by pfmoore** [comment]
Awesome! Thank you so much 🙂

**Comment by PGScully** [comment]
I'm happy to see `leading_powerline_symbol` added.  Should it also work with rprompt, or should there be a `trailing_powerline_symbol`?

**Reply by JanDeDobbeleer** [reply]
Right now it doesn't, but that one could be leveraged to have the same effect. Let me check.

**Reply by JanDeDobbeleer** [reply]
@PGScully this seems to already work just using this configuration:

```json
"powerline_symbol": "\ue0d6",
"leading_powerline_symbol": "\ue0b2",
```

**Reply by PGScully** [reply]
Hmm, I have those two symbols swapped in my config (with `"invert_powerline": true`) and it looks close to how I want it, except the final symbol is a black triangle, not transparent.

My prompt is here: https://github.com/PGScully/omp, and I'm using Windows Terminal (with acrylic background), and Windows is in dark mode.

**Reply by JanDeDobbeleer** [reply]
Did you try the above?

**Reply by PGScully** [reply]
I did, yes.  I eventually got it to work by setting `"invert_powerline": false` only for the rprompt segments using the leading symbol.

**Comment by pfmoore** [comment]
Thinking some more about this problem, while it's now possible to get the sort of effect I was after, I still find it tricky to reason about what's going on, because the "powerline" segment type doesn't really define a start or end character, but rather a *separator* between two segments.

It's probably way too late now to change things, but the way I think about it, segments should just have a "diamond" type ("plain" is simply a diamond with no end characters) and the "powerline" block should be something else - a "section" that contains multiple segments which are separated by the powerline charcater.

So something like:

```
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "blocks": [
    {
      "alignment": "left",
      "segments": [
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part1 "
        },
        {
          "type": "powerline_section",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "separator": "\ue0b0",
          "segments": [
            {
              "type": "text",
              "style": "plain",
              "foreground": "#1c1c1c",
              "background": "#95378f",
              "alias": "A",
              "template": " A "
            },
            {
              "type": "text",
              "style": "plain",
              "foreground": "#1c1c1c",
              "background": "#b925af",
              "alias": "B",
              "template": " B "
            },
            {
              "type": "text",
              "style": "plain",
              "foreground": "#1c1c1c",
              "background": "#df74d8",
              "alias": "C",
              "template": " C "
            }
          ]
        },
        {
          "type": "text",
          "style": "diamond",
          "leading_diamond": "\ue0d7",
          "trailing_diamond": "\ue0b0",
          "foreground": "#1c1c1c",
          "background": "#95378f",
          "template": " Part2 "
        }
      ],
      "type": "prompt"
    }
  ],
  "final_space": true,
  "version": 2
}
```

Segments in a "powerline_section" should probably be required to be type "plain", because diamond doesn't make sense if you're separating the segments with a powerline symbol.

It's a little more manual to get the effect you want with this approach, but there's a lot less context-dependent behaviour, so overall it's (in my view) a lot more understandable. And it works better when you want more complex displays, like

![image](https://github.com/JanDeDobbeleer/oh-my-posh/assets/1110419/198db3e7-a2ea-417c-b799-8bb65d941bb5)

**Reply by JanDeDobbeleer** [reply]
@pfmoore too late for that is indeed a good statement. There's truth to your design, then again what's intuitive for you, might not be for everyone.

**Reply by pfmoore** [reply]
Shame 🙁

I think the thing with powerline is that like many things that "do what you mean" it's great within its specific intended context, but when you try to work beyond that, the implicit behaviour gets in the way too much. I looked at the source, and the `writeSeparator` function is impressively complicated for what it's trying to do 🙂 It's like the issue that started this discussion - working around the Windows Terminal transparency bug is hard, because you end up fighting against some of the built in behaviour of the powerline segment, but you can't just use a diamond segment because you still want *other* parts of what powerline does for you (merging separators when a segment is empty).

Ah well, I think that I'll have to accept a look that isn't quite what I want. It's only cosmetic, after all.

**Reply by JanDeDobbeleer** [reply]
@pfmoore do know that this code is the third iteration of the engine already. It's not easy as feature requests grow over time. I try to rewrite when the time is there, and it's not there yet. It can all be achieved, maybe not as straightforward as you'd like, but it's possible.

**Reply by pfmoore** [reply]
Sorry, I certainly wasn't criticising. My apologies if it came across that way. I know very well how difficult it can be to keep things clean, understandable and maintainable in the face of many feature requests - your responsiveness and willingness to help is really appreciated.

**Reply by JanDeDobbeleer** [reply]
@pfmoore I appreciate the feedback, it's just a lot of work 😅
